### PR TITLE
Proposed fix for issue #358

### PR DESCRIPTION
**Pull request by eachimei**
Switch to `inspect.signature` instead of `inspect.getfullargspec` to properly support wrapper chains (decorated functions) - see details in issue #358 .

In the proposed change I tried to do the minimal required changes to `varnames`.

### CODE REVIEW

**Review by RonnyPfannschmidt — COMMENTED** [review]
Lovely start, please see my comments for the self handling

File: src/pluggy/_hooks.py [hunk 1]
```diff
@@ -245,11 +245,29 @@ def varnames(func: object) -> Tuple[Tuple[str, ...], Tuple[str, ...]]:
             return (), ()
 
     try:  # func MUST be a function or method here or we won't parse any args
-        spec = inspect.getfullargspec(func)
+        sig = inspect.signature(func.__func__ if inspect.ismethod(func) else func)
```
[the remark this answers is not in the document]
Comment (eachimei) [reply src/pluggy/_hooks.py:248]: I thought about that, and the challenge/problem then becomes with cases where `@hookspec` is used on methods or when a class object is passed into `varnames` (then the class `__init__` is used). Meaning, in those cases `func` is an unbound function so `signature(func)` doesn't skip `self`. However, in the existing ("old") implementation, `self` is always stripped (even when `func` is unbound).

File: testing/test_helpers.py [hunk 2]
```diff
@@ -82,3 +84,17 @@ def function4(arg1, *args, **kwargs):
...
+    def my_decorator(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            return func(*args, **kwargs)
+        return wrapper
+
+    @my_decorator
+    def example(a, b=123) -> None:
+        pass
+
```
[the remark this answers is not in the document]
Comment (eachimei) [reply testing/test_helpers.py:101]: added in https://github.com/pytest-dev/pluggy/pull/359/commits/11443b4007895d7d9d560a7aa6bdbe8b6cda243b

File: src/pluggy/_hooks.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (RonnyPfannschmidt) [reply src/pluggy/_hooks.py:248]: those are cases we might want to deprecate in a followup

1. while signatures of `Myclass.__init__` include self, signatures of a `Myclass` do work correct
2. its arguably a mistake that hookspecs as classes where allowed, we should have passed in a object in any case

i'll open up issues for that

**Review by RonnyPfannschmidt — APPROVED** [review]
looks good, thanks

the failures seem unrelated, will fix them later today + merge this

**Comment by eachimei** [comment]
That's great, thank you 😊

**Comment by nicoddemus** [comment]
@eachimei can you please rebase this over `main`? It should run cleanly now.

**Comment by eachimei** [comment]
> @eachimei can you please rebase this over `main`? It should run cleanly now.
@nicoddemus , I rebased so we have fresh results now

**Comment by nicoddemus** [comment]
@RonnyPfannschmidt is this good to merge now?

**Comment by RonnyPfannschmidt** [comment]
It seems like we didn't install pre-commit here as well, the lint fail is from that, so lint fix needed

**Comment by nicoddemus** [comment]
Enabled pre-commit for `pluggy`. 👍

**Comment by nicoddemus** [comment]
I did fix it locally but unfortunately I cannot push to `eachimei:main` (permission denied), here's the diff:

```diff
diff --git a/src/pluggy/_hooks.py b/src/pluggy/_hooks.py
index d376eca..73f0e80 100644
--- a/src/pluggy/_hooks.py
+++ b/src/pluggy/_hooks.py
@@ -244,8 +244,11 @@ def varnames(func: object) -> Tuple[Tuple[str, ...], Tuple[str, ...]]:
         except Exception:
             return (), ()

-    try:  # func MUST be a function or method here or we won't parse any args
-        sig = inspect.signature(func.__func__ if inspect.ismethod(func) else func)
+    try:
+        # func MUST be a function or method here or we won't parse any args.
+        sig = inspect.signature(
+            func.__func__ if inspect.ismethod(func) else func  # type:ignore[arg-type]
+        )
     except TypeError:
         return (), ()

diff --git a/testing/test_helpers.py b/testing/test_helpers.py
index d7ec6ce..4ddd945 100644
--- a/testing/test_helpers.py
+++ b/testing/test_helpers.py
@@ -87,12 +87,13 @@ def test_formatdef() -> None:


 def test_varnames_decorator() -> None:
-    F = TypeVar('F', bound=Callable[..., Any])
+    F = TypeVar("F", bound=Callable[..., Any])

     def my_decorator(func: F) -> F:
         @wraps(func)
         def wrapper(*args, **kwargs):
             return func(*args, **kwargs)
+
         return cast(F, wrapper)

     @my_decorator
```

**Comment by eachimei** [comment]
@nicoddemus can you please retry? It had the "allow edit" setting set to "false", it should work now

**Comment by nicoddemus** [comment]
Yeah now it works. 👍

**Review by RonnyPfannschmidt — APPROVED** [review]

**Pull request merged**